cpu/samd21/periph/adc.c

```diff
@@ -0,0 +1,361 @@
+}
+
+bool adc_syncing(Adc* adc)
+{
+    if(adc->STATUS.reg & ADC_STATUS_SYNCBUSY){
+        return true;
+    }
+    return false;
+}
+
+/* Configure ADC with defined Resolution */
+int adc_configure_with_resolution(Adc* adc, uint32_t precision)
+{
+    adc_poweron(ADC_0);
+    uint8_t  divideResult = ADC_0_DIV_RES_DEFAULT;
+    uint32_t resolution = ADC_0_RES_16BIT;
+    uint32_t accumulate = ADC_0_ACCUM_DEFAULT;
+    if(adc->CTRLA.reg & ADC_CTRLA_SWRST)
+        return -1;
+    if(adc->CTRLA.reg & ADC_CTRLA_ENABLE)
+        return -1;
+    
+    /* GCLK Setup*/
+    GCLK->CLKCTRL.reg = (uint32_t)((GCLK_CLKCTRL_CLKEN
+                          | GCLK_CLKCTRL_GEN_GCLK0 
+                          | (ADC_GCLK_ID << GCLK_CLKCTRL_ID_Pos)));
+
+     /* Pin Muxing */
+    ADC_0_PORT.PINCFG[ ADC_0_POS_INPUT ].bit.PMUXEN = 1;
+    ADC_0_PORT.PMUX[ ADC_0_POS_INPUT / 2].bit.PMUXE = 1;
+
+    /* ADC_0_POS_INPUT Input */
+    ADC_0_PORT.DIRCLR.reg = (1 << ADC_0_POS_INPUT);
+    ADC_0_PORT.PINCFG[ADC_0_POS_INPUT].bit.INEN = true;
+    ADC_0_PORT.PINCFG[ADC_0_POS_INPUT].bit.PULLEN = false; 
+
+    if(ADC_0_NEG_INPUT != ADC_INPUTCTRL_MUXNEG_GND)
+    {
+        ADC_0_PORT.DIRCLR.reg = (1 << ADC_0_NEG_INPUT);
+        ADC_0_PORT.PINCFG[ADC_0_NEG_INPUT].bit.INEN = true;
+        ADC_0_PORT.PINCFG[ADC_0_NEG_INPUT].bit.PULLEN = false;
+    } 
+
+    /* Set RUN_IN_STANDBY */
+    adc->CTRLA.reg = (ADC_0_RUN_IN_STANDBY << ADC_CTRLA_RUNSTDBY_Pos);
+
+    /* Set Voltage Reference */
+    adc->REFCTRL.reg = (ADC_0_REF_COM_EN << ADC_REFCTRL_REFCOMP_Pos) | ADC_0_REF_DEFAULT;
+
+    switch (precision)
+    {
+    case ADC_0_RES_8BIT:
+        resolution = ADC_0_RES_8BIT;
+        break;
+    case ADC_0_RES_10BIT:
+        resolution = ADC_0_RES_10BIT;
+        break;
+    case ADC_0_RES_12BIT:
+        resolution = ADC_0_RES_12BIT;
+        break;
+    case ADC_0_RES_16BIT:
+        divideResult = ADC_0_DIV_RES_DISABLE;
+        switch(ADC_0_ACCUM_DEFAULT)
+        {
+            case ADC_0_ACCUM_512:
+                accumulate = ADC_0_ACCUM_512;
+                break;
+            case ADC_0_ACCUM_1024:
+                accumulate = ADC_0_ACCUM_1024;
+                break;
+            default:
+                accumulate = ADC_0_ACCUM_256;
+                break;
+        }        
+        resolution = ADC_0_RES_16BIT;
+        break;
+    default:
+        DEBUG("Invalid precision");        
+        return -1;
+    }
+
+    /* Set the accumlation and devide result */
+    adc->AVGCTRL.reg = ADC_AVGCTRL_ADJRES(divideResult) | accumulate;
+
+    /* Set Sample length */
+    adc->SAMPCTRL.reg = (ADC_0_SAMPLE_LENGTH << ADC_SAMPCTRL_SAMPLEN_Pos);
+    while(adc_syncing(adc));
+    
+    /* If external vref. Pin setup */
+    if(ADC_0_REF_DEFAULT == ADC_0_REF_EXT_B)
+    {
+        ADC_0_PORT.DIRCLR.reg = (1 << ADC_0_REF_DEFAULT);
+        ADC_0_PORT.PINCFG[ADC_0_REF_DEFAULT].bit.INEN = true;
+        ADC_0_PORT.PINCFG[ADC_0_REF_DEFAULT].bit.PULLEN = false;
+    }
```

Comment:
Don't you need to set up the `REFCTRL` register (sec. 13.8.2 p.818) for the external referenc pin `VREFB`?